### PR TITLE
[SPARK-33187][SQL] Add a check on the number of returned partitions in the HiveShim#getPartitionsByFilter method

### PR DESCRIPTION
**Pull request by jinhai-cloud**
### What changes were proposed in this pull request?

Add a check on the number of returned metastore partitions by calling Hive#getNumPartitionsByFilter, and add SQLConf spark.sql.hive.metastorePartitionLimit, default value is 100_000

### Why are the changes needed?

In the method Shim#getPartitionsByFilter, when filter is empty or when the hive table has a large number of partitions, calling getAllPartitionsMethod or getPartitionsByFilterMethod will results in Driver OOM.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

This change is already covered by existing tests

### CODE REVIEW

**Comment by manbuyun** [comment]
retest this please

**Comment by manbuyun** [comment]
@HyukjinKwon Can you help verify this patch?

**Comment by maropu** [comment]
Could you add tests? btw, does hive have a similar config?

**Comment by wangyum** [comment]
Is it a good design if partition number larger than 100000?

**Comment by manbuyun** [comment]
> Is it a good design if partition number larger than 100000?

User queries generally do not exceed 100000 partitions. It's just that filter can't filter partitions very well.
For example, the expression is of a different type, issue: https://issues.apache.org/jira/browse/SPARK-15287.
And Presto also has a parameter to limit the number of partitions in HiveMetadata#getPartitionsAsList

image

**Comment by manbuyun** [comment]
> Could you add tests? btw, does hive have a similar config?

I'll think about adding a test. 
HiveMetaStore#get_partitions_by_filter also call method checkLimitNumberOfPartitionsByFilter，and HiveConf#METASTORE_LIMIT_PARTITION_REQUEST default value is -1.
image
image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -815,6 +815,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val HIVE_METASTORE_PARTITION_LIMIT =
+    buildConf("spark.sql.hive.metastorePartitionLimit")
+      .doc("The maximum number of metastore partitions allowed. Use -1 for unlimited.")
+      .version("3.0.2")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:821]: `3.0.2` -> `3.1.0`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -815,6 +815,14 @@ object SQLConf {
...
+      .intConf
+      .checkValue(_ >= -1, "The maximum must be a positive integer, -1 to apply no limit.")
+      .createWithDefault(100000)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:824]: `-1` to follow the Hive config?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 3]
```diff
@@ -1233,6 +1242,34 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
...
+    // Hive getPartitionsByFilter() takes a string that represents partition
+    // predicates like "str_key=\"value\" and int_key=1 ..."
+    val filter = convertFilters(table, predicates)
+
+    val limit = SQLConf.get.metastorePartitionLimit
+    if (limit > -1) {
+      val num = try {
+        getNumPartitionsByFilterMethod.invoke(hive, table, filter).asInstanceOf[Int]
+      } catch {
+        case ex: Exception =>
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:1260]: `case NonFatal(_) `?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 4]
```diff
@@ -1233,6 +1242,34 @@ private[client] class Shim_v2_1 extends Shim_v2_0 {
...
+          logWarning("Caught Hive MetaException attempting to get partition metadata by " +
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:1261]: Could you make the message clearer? e.g., ...get the number of partitions from ...., but .... using 0 for a partition number....

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 5]
```diff
@@ -773,26 +773,7 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
     filters.flatMap(convert).mkString(" and ")
   }
 
-  private def quoteStringLiteral(str: String): String = {
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:776]: Why did you move this func?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -815,6 +815,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val HIVE_METASTORE_PARTITION_LIMIT =
+    buildConf("spark.sql.hive.metastorePartitionLimit")
+      .doc("The maximum number of metastore partitions allowed. Use -1 for unlimited.")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:820]: The current statement looks a bit ambiguous, so how about following the statement of the Hive config you pointed out?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 5 above)
Comment (manbuyun) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:776]: Didn't remove this method, just added the getPartitionsByFilterInternal method

**Review by wangyum — COMMENTED** [review]
Could you verify is `hive.metastore.limit.partition.request` works? if it works, we can add it to document. We have added a [Hadoop config](https://github.com/apache/spark/pull/17442/files) before: `mapreduce.fileoutputcommitter.algorithm.version`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 7]
```diff
@@ -815,6 +815,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val HIVE_METASTORE_PARTITION_LIMIT =
+    buildConf("spark.sql.hive.metastorePartitionLimit")
+      .doc("The maximum number of metastore partitions allowed for a given table. The default " +
+           "value -1 to follow the Hive config (see HiveConf.METASTORE_LIMIT_PARTITION_REQUEST " +
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:821]: nit: can we use `hive.metastore.limit.partition.request instead of `HiveConf.METASTORE_LIMIT_PARTITION_REQUEST` here? also do we want to mention this is only used for HMS version 2.1 and up?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 8]
```diff
@@ -815,6 +815,16 @@ object SQLConf {
...
+           "for more information).")
+      .version("3.1.0")
+      .intConf
+      .checkValue(_ >= -1, "The maximum must be a positive integer, -1 to follow the Hive config.")
+      .createWithDefault(100000)
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:826]: have you considered to keep the default behavior as it is but allow it to be configurable? changing it means now we'll need to make two HMS calls (one additional `getNumPartitionsByFilter`) which I'm not sure is desirable (have seen HMS perform very badly in production before).

**Comment by manbuyun** [comment]
> Could you verify is `hive.metastore.limit.partition.request` works? if it works, we can add it to document. We have added a [Hadoop config](https://github.com/apache/spark/pull/17442/files) before: `mapreduce.fileoutputcommitter.algorithm.version`.

It does not work on the client

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (manbuyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:826]: @sunchao Thank you for your response. I think this is a reasonable maximum, and Presto also has a parameter to limit the number of partitions in HiveMetadata#getPartitionsAsList, default value is 100_000

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:847]: Yea the default value 100_000 looks fine to me. My main question is whether we need to make the default value to be that and double the HMS calls. Presto doesn't call `getNumPartitionsByFilter` it seems as it streams through a partition iterator and stops once the threshold is reached.

**Comment by manbuyun** [comment]
@sunchao @wangyum Can you help me verify this patch?
